cli/src/complete.rs

```diff
@@ -354,6 +354,244 @@ pub fn leaf_config_keys() -> Vec<CompletionCandidate> {
+    }
+
+    fn parse_range_impl<T>(args: impl Fn() -> T) -> Option<(String, String)>
+    where
+        T: Iterator<Item = String>,
+    {
+        let from = parse_flag(&["-f", "--from"], args())?;
+        let to = parse_flag(&["-t", "--to"], args()).unwrap_or_else(|| "@".into());
+
+        Some((from, to))
+    }
+
+    #[test]
+    fn test_parse_range_impl() {
+        let wc_cases: &[&[&str]] = &[
+            &["-f", "foo"],
+            &["--from", "foo"],
+            &["-f=foo"],
+            &["preceding_arg", "-f", "foo"],
+            &["-f", "foo", "following_arg"],
+        ];
+        for case in wc_cases {
+            let args = case.iter().map(|s| s.to_string());
+            assert_eq!(
+                parse_range_impl(|| args.clone()),
+                Some(("foo".into(), "@".into())),
+                "case: {case:?}",
+            );
+        }
+        let to_cases: &[&[&str]] = &[
+            &["-f", "foo", "-t", "bar"],
+            &["-f", "foo", "--to", "bar"],
+            &["-f=foo", "-t=bar"],
+            &["-t=bar", "-f=foo"],
+        ];
+        for case in to_cases {
+            let args = case.iter().map(|s| s.to_string());
+            assert_eq!(
+                parse_range_impl(|| args.clone()),
+                Some(("foo".into(), "bar".into())),
+                "case: {case:?}",
+            );
+        }
+        let bad_cases: &[&[&str]] = &[&[], &["-f"], &["foo"], &["-R", "foo"], &["-R=foo"]];
+        for case in bad_cases {
+            let args = case.iter().map(|s| s.to_string());
+            assert_eq!(parse_range_impl(|| args.clone()), None, "case: {case:?}");
+        }
+    }
+
+    pub fn range() -> Option<(String, String)> {
+        parse_range_impl(std::env::args)
+    }
+
+    // Special parse function only for `jj squash`. While squash has --from and
+    // --to arguments, only files within --from should be completed, because
+    // the files changed only in some other revision in the range between
+    // --from and --to cannot be squashed into --to like that.
+    pub fn squash_revision() -> Option<String> {
+        if let Some(rev) = parse_flag(&["-r", "--revision"], std::env::args()) {
+            return Some(rev);
+        }
+        parse_flag(&["-f", "--from"], std::env::args())
+    }
+}
+
+fn all_files_from_rev(rev: String) -> Vec<CompletionCandidate> {
+    with_jj(|mut jj, _| {
+        let output = jj
+            .arg("file")
+            .arg("list")
+            .arg("--revision")
+            .arg(rev)
+            .output()
+            .map_err(user_error)?;
+        let stdout = String::from_utf8_lossy(&output.stdout);
+
+        Ok(stdout.lines().map(CompletionCandidate::new).collect())
+    })
+}
+
+fn modified_files_from_rev(
+    rev: (String, Option<String>),
+    interdiff: bool,
+) -> Vec<CompletionCandidate> {
+    with_jj(|mut jj, _| {
+        let cmd = jj
+            .arg(if interdiff { "interdiff" } else { "diff" })
```

Comment:
I wouldn't expect `jj interdiff ... <TAB>` does a temporary rebase to provide candidates. I'd rather want a completion that is lightweight and includes some false positives.

Reply:
I made it so all files in `from` and `to` are included, that seems reasonable. Adding the `JjBuilder` (or something similar) should also help [here](https://github.com/martinvonz/jj/pull/4900#discussion_r1848109413).